plugins/plugin-factory-github/codenvy-plugin-factory-github-server/src/main/java/com/codenvy/plugin/factory/github/server/GithubFactoryParametersResolver.java

```diff
@@ -0,0 +1,194 @@
+
+        // use master if not specified
+        if (branchName == null) {
+            branchName = "master";
+        }
+
+        // check if repository contains a codenvy Dockerfile
+        String dockerFileLocation = null;
+        String githubDockerFileLocation =
+                "https://raw.githubusercontent.com/".concat(repoUser).concat("/").concat(repoName).concat("/").concat(branchName)
+                                                    .concat("/.codenvy.dockerfile");
+        HttpURLConnection httpURLConnection = null;
+        try {
+            httpURLConnection = (HttpURLConnection)new URL(githubDockerFileLocation).openConnection();
+            httpURLConnection.setConnectTimeout(CONNECTION_TIMEOUT);
+            final int responseCode = httpURLConnection.getResponseCode();
+            if (responseCode == HttpURLConnection.HTTP_OK) {
+                dockerFileLocation = githubDockerFileLocation;
+            }
+        } catch (IOException ioe) {
+            LOG.debug("Unable to check if remote location {0} is available or not", githubDockerFileLocation);
+        } finally {
+            if (httpURLConnection != null) {
+                httpURLConnection.disconnect();
+            }
+        }
+
+        // apply default location if there is none
+        if (dockerFileLocation == null) {
+            dockerFileLocation = DEFAULT_DOCKER_FILE_LOCATION;
+        }
+
+        String envName = repoUser;
+
+        SourceStorageDto
+                sourceStorageDto = newDto(SourceStorageDto.class).withLocation(githubLocation).withType("git").withParameters(parameters);
+        ProjectConfigDto projectConfig = newDto(ProjectConfigDto.class).withSource(sourceStorageDto).withName(repoName).withType("blank")
+                                                                       .withPath("/".concat(repoName));
+
+        LimitsDto limitsDto = newDto(LimitsDto.class).withRam(2000);
+        MachineSourceDto machineSourceDto = newDto(MachineSourceDto.class).withType("dockerfile").withLocation(dockerFileLocation);
+        MachineConfigDto
+                machineConfigDto =
+                newDto(MachineConfigDto.class).withLimits(limitsDto).withType("docker").withSource(machineSourceDto).withDev(true)
+                                              .withName("ws-machine");
+
+        EnvironmentDto environmentDto = newDto(EnvironmentDto.class).withName(envName).withMachineConfigs(singletonList(machineConfigDto));
+
+        final WorkspaceConfigDto
+                workspaceConfig = newDto(WorkspaceConfigDto.class).withDefaultEnv(envName).withProjects(singletonList(projectConfig))
+                                                                  .withEnvironments(singletonList(environmentDto)).withName(repoUser);
+        final Factory factory = newDto(Factory.class).withWorkspace(workspaceConfig).withV("4.0");
+
+        return factory;
```

Comment:
but AFAIK there is backward compliance in the model as we have a version in the Factory object (v 4.0, etc) so there shouldn't have any manual changes to do.
So it means that if attributes are changed, version number is bumped. but the DTO is still valid.
As the factory.json file can be stored on client side or on repository it should always be valid.


The only point I see is introducing a default factory builder to create factory with default parameters and only keep github specific part there